### ISSUE TITLE
brcmfmac direct firmware load failed

### ISSUE THREAD

**Issue by Arkham** [nerves-project/nerves_system_rpi3#8]
Hi there,

I'm trying to get wifi working on my raspberry pi 3, and I randomly get these errors on boot:

```
[    3.058626] brcmfmac: brcmf_sdio_drivestrengthinit: No SDIO Drive strength init done for chip 43430 rev 1 pmurev 24
[    3.059012] brcmfmac_sdio mmc1:0001:1: Direct firmware load for brcm/brcmfmac43430-sdio.bin failed with error -2
[    4.077545] brcmfmac: brcmf_sdio_htclk: HT Avail timeout (1000000): clkctl 0x50
[    5.088454] brcmfmac: brcmf_sdio_htclk: HT Avail timeout (1000000): clkctl 0x50
```

After that I don't see wlan0 among the available interfaces. Here are the steps I've followed.

1. Get latest version of nerves-bootstrap `0.2.1`
2. Run `mix new hello_nerves --target rpi3` which installs version `0.9.1` of nerves_system_rpi3
3. Add `nerves_interim_wifi` and set it up on application start

Here's my mix.lock
```
%{"distillery": {:hex, :distillery, "1.0.0", "a866a72bf2a3a5f078f5a249017ed951acda88a760d200512f91f585d74db1ec", [:mix], []},
  "elixir_make": {:hex, :elixir_make, "0.4.0", "992f38fabe705bb45821a728f20914c554b276838433349d4f2341f7a687cddf", [:mix], []},
  "nerves": {:hex, :nerves, "0.4.5", "1a82415b7b60fef675a50be85ba8c37b13713d45a47acdded64e6e28ea9309a5", [:mix], [{:distillery, "~> 1.0.0", [hex: :distillery, optional: false]}]},
  "nerves_interim_wifi": {:hex, :nerves_interim_wifi, "0.1.0", "f572381722a06f1cbc118708baecd1b12ce00edb684d88258aa417112f32489a", [:make, :mix], [{:elixir_make, "~> 0.3", [hex: :elixir_make, optional: false]}, {:nerves_network_interface, "~> 0.3.1", [hex: :nerves_network_interface, optional: false]}, {:nerves_wpa_supplicant, "~> 0.2.2", [hex: :nerves_wpa_supplicant, optional: false]}]},
  "nerves_network_interface": {:hex, :nerves_network_interface, "0.3.2", "e7450f7eccd7387ceebee6096a0646bcdc0dd4aa77077e5fa81f6767da9b77af", [:make, :mix], [{:elixir_make, "~> 0.3", [hex: :elixir_make, optional: false]}]},
  "nerves_system_br": {:hex, :nerves_system_br, "0.8.2", "f9917e75b54fb3eadbd67c2123ac1406d0a7fdef24f1960e3577c759d0a6fbc7", [:mix], []},
  "nerves_system_rpi3": {:hex, :nerves_system_rpi3, "0.9.1", "4cccb5ec4f34c1dafea1ee613ffefd31db82da545ac37cefb197da5c804a235b", [:mix], [{:nerves, "~> 0.4.0", [hex: :nerves, optional: false]}, {:nerves_system_br, "~> 0.8.1", [hex: :nerves_system_br, optional: false]}, {:nerves_toolchain_arm_unknown_linux_gnueabihf, "~> 0.8.0", [hex: :nerves_toolchain_arm_unknown_linux_gnueabihf, optional: false]}]},
  "nerves_toolchain_arm_unknown_linux_gnueabihf": {:hex, :nerves_toolchain_arm_unknown_linux_gnueabihf, "0.8.0", "873f8d6f74413c14eda3462b397a88f043a6c3bd65f56ff57db4e2e1eaf48994", [:mix], [{:nerves, "~> 0.4.0", [hex: :nerves, optional: false]}, {:nerves_toolchain_ctng, "~> 0.8.0", [hex: :nerves_toolchain_ctng, optional: false]}]},
  "nerves_toolchain_ctng": {:hex, :nerves_toolchain_ctng, "0.8.0", "6dff7ed51e1711c5f4da3d559bc528a8265e3dd950dda95f4d6832aed9dbe320", [:mix], []},
  "nerves_wpa_supplicant": {:hex, :nerves_wpa_supplicant, "0.2.2", "da829baec94a8afa7da3a6beaaf1725b9ca7f5f58779ba50b73568b7d37d3d98", [:make, :mix], [{:elixir_make, "~> 0.3", [hex: :elixir_make, optional: false]}]}}
```

**Comment by fhunleth** [comment]
Could you try adding all of the files in this directory https://github.com/RPi-Distro/firmware-nonfree/tree/master/brcm80211/brcm (and in particular, `brcmfmac43430-sdio.bin`) to your `rootfs-additions/lib/firmware/brcm` directory in your project?

The `brcmfmac43430-sdio.bin` file had been recently added in #6, but I didn't know the official source. Since I now know the official source, I'm thinking of dumping all of the firmware in the `brcm` directory above into our repo. We have a release pending, so if you could verify that those files work for you, I'll add them right away. Amazingly enough the Raspberry Pi 3s that I have don't exhibit this problem.

**Comment by Arkham** [comment]
Hey @fhunleth, I've tried replacing the brcm firmware with all the variants I could find on the internet :) So I've tried with the ones available in the repo you posted and then I found a slightly newer version here https://github.com/OpenELEC/wlan-firmware/tree/master/firmware/brcm.

Even with the updated drivers, the results are fluctuating quite a lot: when I keep rebooting the PI the firmware will load correctly only 25% of the times. Could you could please post somewhere a working repo so that I can try it on my Pi? Also I've pushed the repo I've been testing stuff on here https://github.com/Arkham/nerves_rpi3_wifi_test

**Comment by fhunleth** [comment]
Hmm. Have you tried copying all of the files in the brcm directory onto your Pi? Maybe the error message is bad and it fails to load since it can't find one of the other files in the directory. That 25% number is really weird, since that indicates that the files are there. My next guess is that the WiFi driver prefers to be loaded as a module or that it only loads 1/4 the time and Raspbian just retries until it works. 

Do you know how to recompile the WiFi driver as a module and then `modprobe` it from Elixir?

Sorry to not be more help - my RPi3 doesn't require `brcmfmac43430-sdio.bin`, so it appears that they changed the WiFi chips on the RPi3's at some point. I know where I can borrow a new RPi3, but I can't get to trying it out until probably next week.

**Comment by Arkham** [comment]
Searching a bit on the internet it seems that people are doing the `modprobe -r <module> && modprobe <module>` to get the thing to work. I don't really know how to recompile the driver myself though, could you point me to a guide to do that? Thanks!

Also I've tried this on another RPi3 I managed to borrow, but unfortunately the issue is present there as well.

**Comment by fhunleth** [comment]
You'll need to create a custom system. This is actually a bit of work if you haven't done it before, but here are the docs: https://hexdocs.pm/nerves/systems.html#creating-or-modifying-a-nerves-system-with-buildroot.

Run `make linux-menuconfig` and navigate to "Device Drivers->Network device support->Wireless LAN". Look for "Broadcom IEEE802.11n embedded FullMAC WLAN driver" and type "m". Then keep going to "Exit" until it really exits. You probably should run `make linux-savedefconfig` and copy the `build/linux-xxx/defconfig` over to the system's `linux-4.4.defconfig` to save the change. If you do a `git diff`, you should see that the only change was for the Broadcom wifi driver. From there, run `make` to build the system image. and follow the hexdocs.

**Comment by Arkham** [comment]
Thank you so much, I will give it a try this weekend :)

**Comment by fhunleth** [comment]
While talking to @mobileoverlord, we realized that the Raspberry Pi kernels that we use is old. Well, they're not that old, but they pre-date the time when we started hearing reports about RPi3 wifi issues. I've updated them from Linux kernel 4.4.8 to 4.4.43. Unfortunately, the rpi-firmware package is tightly coupled to the kernel, so I updated that too. And then I updated rpi-userland just because I always update them all together. The net of this is that we'll need to release an updated `nerves_system_br` with the package updates.

I'm currently building the updates to the official RPi system images. My hope is that someone upstream from us has fixed the WiFi issue. Anyway, I wouldn't do any testing unless your using the 4.4.43 kernel. Hopefully, we can have that out by the weekend.

**Comment by Arkham** [comment]
Thanks so much, really appreciate it! I will gladly test out the updated `nerves_system_br` as soon as it comes out. Let me know if I can do anything to help out as well :)

**Comment by Arkham** [comment]
The problem has been fixed with the lates `0.10.0` release. Thanks! 🎊

**Comment by mleonhard** [comment]
What was the final fix?  I'm getting a similar error with nixos on rpi4.

```
brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac43455-sdio.bin failed with error -2
```
